roles/v2/pool/src/lib/mining_pool/mod.rs

```diff
@@ -523,6 +537,11 @@ impl Pool {
                 error!("Downstream shutdown and Status Channel dropped");
             }
         });
+        cloned3
+    }
+
+    pub fn drop_downstream(&mut self, downstream_id: u32) {
```

Comment:
Why use a function if the only thing that we do is calling remove? Also Arc<Downstream> can be cloned are we sure that removing it from the self.downstream HashMap is sufficient for actually drop every reference to downstream? With the code as it today seems so but maybe in the future will not be anymore the case, maybe (if possible) we should add a test that make sure that drop downstream remove every reverence to downstream.

Possibe thing that could work just thinking out loud:
```rust
    pub fn drop_downstream(&mut self, downstream_id: u32) {
        let d = self.downstreams.remove(&downstream_id).unwrap();
        // we could fail early in debug enviornment this do not assure anything btw
        debug_assert!(Arc::strong_count(&d) == 1);
        // or we could drop the value inside the safe lock not sure what will happen
        d.safe_lock(|d| drop(d)).unwrap();
    }
```

Reply:
The drop_downstream is used to access the downstreams hashmap which isn't public. This makes sense to me. Here is the bigger issue that I found/fixed:

If a downstream gets dropped it gets removed from the downstreams list... but because we clone the downstreams there is a race condition like this:

 

1.) New template comes in and we clone the downstreams with 100 downstreams

2.) downstream #6 disconnects and is removed from downstreams

3.) as we are iterating the clone of downstreams from step 1 downstream #6 still exists in the clone

4.) we hit #6 and try to send the template to them - this fails as they aren't actually there anymore

5.) the handle_result (i think) is a break on this error and so the iteration doesn't complete and downstreams 7-100 never get their new templates

Reply:
Don't you think that we could have memory leaks? What worry me is that in some future version of the code someone call `drop_dowstream` and think to drop it but maybe some clone of the `Arc` is still around. Btw I'm ok of merging it as it is maybe just change the function name or put the `debug_assert`. What do you think?

Reply:
There is no way to guarantee that a future dev won't introduce memory leaks. This additional way to drop downstreams doesn't make the opportunity for memory leaks greater. I'm also fine with adding the debug_assert but I don't think it adds any value - it's perfectly reasonable for a downstream to be removed from the main version of the map but still reside (temporarily) in a cloned version. Currently all the cloned versions are only in scope for a short amount of time and once they go out of scope they are freed.

Reply:
agree what worry me is the name if I call drop_something I expect to drop the actual thing not just the Arc. The debug assert IMHO in the case of a future dev that call drop thinking of dropping downstream and not just the Arc will be maybe notified during the tests that the assumption is not correct, not much value btw.

Reply:
If you wanted to drop the actual thing, could you expose the the actual thing's mutex guard via the mutex wrapper's `to_remove()`?